# UTS 

<table>
  <tr>
    <td>Nama</td>
    <td>:</td>
    <td>Ahmad Farisul Haq</td>
  </tr>
  <tr>
    <td>NIM</td>
    <td>:</td>
    <td>200411100171</td>
  </tr>
  <tr>
    <td>Kelas</td>
    <td>:</td>
    <td>Penambangan Data A</td>
  </tr>
</table>

## Implementasi dengan K-NN

In [158]:
import pandas as pd
import numpy as np

In [159]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/00451/dataR2.csv'

df = pd.read_csv(url)
df

Age        BMI  Glucose  Insulin      HOMA   Leptin  Adiponectin  \
0     48  23.500000       70    2.707  0.467409   8.8071     9.702400   
1     83  20.690495       92    3.115  0.706897   8.8438     5.429285   
2     82  23.124670       91    4.498  1.009651  17.9393    22.432040   
3     68  21.367521       77    3.226  0.612725   9.8827     7.169560   
4     86  21.111111       92    3.549  0.805386   6.6994     4.819240   
..   ...        ...      ...      ...       ...      ...          ...   
111   45  26.850000       92    3.330  0.755688  54.6800    12.100000   
112   62  26.840000      100    4.530  1.117400  12.4500    21.420000   
113   65  32.050000       97    5.730  1.370998  61.4800    22.540000   
114   72  25.590000       82    2.820  0.570392  24.9600    33.750000   
115   86  27.180000      138   19.910  6.777364  90.2800    14.110000   

     Resistin    MCP.1  Classification  
0     7.99585  417.114               1  
1     4.06405  468.786               1  
2     9.27715  554.697               1  
3    12.76600  928.220               1  
4    10.57635  773.920               1  
..        ...      ...             ...  
111  10.96000  268.230               2  
112   7.32000  330.160               2  
113  10.33000  314.050               2  
114   3.27000  392.460               2  
115   4.35000   90.090               2  

[116 rows x 10 columns]

### cek ukuran dataset 

In [160]:
df.shape

(116, 10)

### Cek missing value dataset

In [161]:
df.isna().sum()

Age               0
BMI               0
Glucose           0
Insulin           0
HOMA              0
Leptin            0
Adiponectin       0
Resistin          0
MCP.1             0
Classification    0
dtype: int64

### data analysis

In [162]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 116 entries, 0 to 115
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             116 non-null    int64  
 1   BMI             116 non-null    float64
 2   Glucose         116 non-null    int64  
 3   Insulin         116 non-null    float64
 4   HOMA            116 non-null    float64
 5   Leptin          116 non-null    float64
 6   Adiponectin     116 non-null    float64
 7   Resistin        116 non-null    float64
 8   MCP.1           116 non-null    float64
 9   Classification  116 non-null    int64  
dtypes: float64(7), int64(3)
memory usage: 9.2 KB


### memisah dataset

In [163]:
X = df.iloc[:, :9]
y = df.iloc[:, 9]

X, y

(     Age        BMI  Glucose  Insulin      HOMA   Leptin  Adiponectin  \
 0     48  23.500000       70    2.707  0.467409   8.8071     9.702400   
 1     83  20.690495       92    3.115  0.706897   8.8438     5.429285   
 2     82  23.124670       91    4.498  1.009651  17.9393    22.432040   
 3     68  21.367521       77    3.226  0.612725   9.8827     7.169560   
 4     86  21.111111       92    3.549  0.805386   6.6994     4.819240   
 ..   ...        ...      ...      ...       ...      ...          ...   
 111   45  26.850000       92    3.330  0.755688  54.6800    12.100000   
 112   62  26.840000      100    4.530  1.117400  12.4500    21.420000   
 113   65  32.050000       97    5.730  1.370998  61.4800    22.540000   
 114   72  25.590000       82    2.820  0.570392  24.9600    33.750000   
 115   86  27.180000      138   19.910  6.777364  90.2800    14.110000   
 
      Resistin    MCP.1  
 0     7.99585  417.114  
 1     4.06405  468.786  
 2     9.27715  554.697  
 3    

### preprocessing data

In [164]:
from sklearn.preprocessing import StandardScaler
ss=StandardScaler()
X=ss.fit_transform(X)
X=pd.DataFrame(X)
X

0         1         2         3         4         5         6  \
0   -0.579794 -0.816675 -1.239222 -0.728739 -0.614282 -0.932334 -0.070222   
1    1.601821 -1.378751 -0.258299 -0.688038 -0.548240 -0.930413 -0.697350   
2    1.539489 -0.891764 -0.302887 -0.550073 -0.464752 -0.454219  1.797998   
3    0.666843 -1.243303 -0.927110 -0.676965 -0.574210 -0.876021 -0.441945   
4    1.788816 -1.294601 -0.258299 -0.644743 -0.521081 -1.042682 -0.786881   
..        ...       ...       ...       ...       ...       ...       ...   
111 -0.766789 -0.146468 -0.258299 -0.666590 -0.534786  1.469335  0.281654   
112  0.292852 -0.148468  0.098400 -0.546881 -0.435039 -0.741611  1.649470   
113  0.479848  0.893855 -0.035362 -0.427172 -0.365106  1.825348  1.813843   
114  0.916171 -0.398546 -0.704173 -0.717467 -0.585883 -0.086651  3.459038   
115  1.788816 -0.080447  1.792721  0.987394  1.125766  3.333167  0.576644   

            7         8  
0   -0.545517 -0.341251  
1   -0.864214 -0.191224  
2   -0.441660  0.058214  
3   -0.158867  1.142718  
4   -0.336352  0.694716  
..        ...       ...  
111 -0.305255 -0.773527  
112 -0.600299 -0.593717  
113 -0.356320 -0.640491  
114 -0.928577 -0.412832  
115 -0.841036 -1.290747  

[116 rows x 9 columns]

### menyemplit dataset dengan perbandingan data train 80% dan data tes 20%

In [165]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test=train_test_split(X, y, test_size=0.2, random_state=1)

In [166]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((92, 9), (24, 9), (92,), (24,))

### Model K-NN dengan sklearn

In [167]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report

my_param_grid = {'n_neighbors':[2,3,5,7], 'weights': ['distance', 'uniform']}
GridSearchCV(estimator=KNeighborsClassifier(), param_grid=my_param_grid, refit=True, verbose=3, cv=3)
modelKNN = GridSearchCV(KNeighborsClassifier(), param_grid=my_param_grid, refit=True, verbose=3, cv=3)


In [168]:
my_param_grid = {'n_neighbors':[2,3,5,7], 'weights': ['distance', 'uniform']}
GridSearchCV(estimator=KNeighborsClassifier(), param_grid=my_param_grid, refit=True, verbose=3, cv=3)
knn = GridSearchCV(KNeighborsClassifier(), param_grid=my_param_grid, refit=True, verbose=3, cv=3)

In [169]:
knn.fit(X_train, y_train)

Fitting 3 folds for each of 8 candidates, totalling 24 fits
[CV 1/3] END ...n_neighbors=2, weights=distance;, score=0.645 total time=   0.0s
[CV 2/3] END ...n_neighbors=2, weights=distance;, score=0.806 total time=   0.0s
[CV 3/3] END ...n_neighbors=2, weights=distance;, score=0.700 total time=   0.0s
[CV 1/3] END ....n_neighbors=2, weights=uniform;, score=0.677 total time=   0.0s
[CV 2/3] END ....n_neighbors=2, weights=uniform;, score=0.710 total time=   0.0s
[CV 3/3] END ....n_neighbors=2, weights=uniform;, score=0.767 total time=   0.0s
[CV 1/3] END ...n_neighbors=3, weights=distance;, score=0.742 total time=   0.0s
[CV 2/3] END ...n_neighbors=3, weights=distance;, score=0.710 total time=   0.0s
[CV 3/3] END ...n_neighbors=3, weights=distance;, score=0.733 total time=   0.0s
[CV 1/3] END ....n_neighbors=3, weights=uniform;, score=0.742 total time=   0.0s
[CV 2/3] END ....n_neighbors=3, weights=uniform;, score=0.710 total time=   0.0s
[CV 3/3] END ....n_neighbors=3, weights=uniform;,

GridSearchCV(cv=3, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [2, 3, 5, 7],
                         'weights': ['distance', 'uniform']},
             verbose=3)

In [170]:
knn.best_params_

{'n_neighbors': 3, 'weights': 'distance'}

### karena nilai k terbaik dari uji coba yaitu k = 2, 3,5,7 maka kita bisa gunakan {'n_neighbors': 3, 'weights': 'distance'}

### data test predict

In [171]:
test_pred = knn.predict(X_test)

### akurasi 

In [172]:
print(classification_report(y_test, test_pred))

              precision    recall  f1-score   support

           1       0.57      0.67      0.62         6
           2       0.88      0.83      0.86        18

    accuracy                           0.79        24
   macro avg       0.73      0.75      0.74        24
weighted avg       0.80      0.79      0.80        24



### inputan test

In [173]:
y_pred = np.array([1.788816,	-0.080447,	1.792721,	0.987394,	1.125766,	3.333167,	0.576644,	-0.841036,	-1.290747,]).reshape(1,-1)
y_pred.shape

(1, 9)

### melakukan predict terhadap inputan

In [174]:
y_predict = knn.predict(y_pred)

In [175]:
labels = ['Healthy controls', 'Patients']

In [176]:
labels[y_predict[0]-1]

'Patients'

## Implementasi dengan Desision tree

### Model Desision tree dengan sklearn

In [177]:
#Using DecisionTreeClassifier of tree class to use Decision Tree Algorithm
from sklearn.tree import DecisionTreeClassifier
tree =DecisionTreeClassifier(criterion = 'entropy', random_state = 1)
tree.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy', random_state=1)

### predict data test

In [178]:
y_pred = tree.predict(X_test)

### akurasi 

In [179]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           1       0.67      0.67      0.67         6
           2       0.89      0.89      0.89        18

    accuracy                           0.83        24
   macro avg       0.78      0.78      0.78        24
weighted avg       0.83      0.83      0.83        24



### inputan test

In [180]:
y_pred = np.array([1.788816,	-0.080447,	1.792721,	0.987394,	1.125766,	3.333167,	0.576644,	-0.841036,	-1.290747,]).reshape(1,-1)
y_pred.shape

(1, 9)

### melakukan predict terhadap inputan

In [181]:
y_predic = tree.predict(y_pred)

In [182]:
labels[y_predict[0]-1]

'Patients'

## Kesimpulan

Dapat disimpulakan bahwa pada dataset Coimbra Breast Cancer antara metode K-NN dan Desision tree mengahsilkan akurasi yang berbeda, dari akurasi K-NN tertinggi 79% dan Desision tree 83%